### PR TITLE
Fixes #5999 - Support assigning public IP to Amazon EC2 host

### PR DESCRIPTION
**Pull request by haf**
This is a shot in the dark, hopefully it hits the mark.

/cc @karaaie @KrisBuytaert

### CODE REVIEW

**Comment by KrisBuytaert** [comment]
This PR should fix http://projects.theforeman.org/issues/5999

File: app/models/compute_resources/foreman/model/ec2.rb [hunk 1]
```diff
@@ -45,6 +45,7 @@ def create_vm(args = { })
       end
       args[:groups].reject!(&:empty?) if args.has_key?(:groups)
       args[:security_group_ids].reject!(&:empty?) if args.has_key?(:security_group_ids)
+      args[:associate_public_ip] = true if vpc && vm.ip_address == 'public'
```
Comment (dLobatog) [inline app/models/compute_resources/foreman/model/ec2.rb]: Looks good, but `vpc` is not defined?

**Comment by dLobatog** [comment]
[test]

**Comment by dLobatog** [comment]
@haf Looks good, except for that issue about `vpc`. Could you recommit following the commit conventions explained above by @theforeman-bot ?

Another thing that caught me off guard is the Vagrantfile, was that included by mistake?

**Comment by haf** [comment]
Hi, yes, I can rebase and such; the hard part here is that I can't easily get [my environment](http://projects.theforeman.org/issues/8394) up to test it.

As for the vagrant file, I should remove it.

**Comment by haf** [comment]
How do I check if we're in a vpc?

File: app/models/compute_resources/foreman/model/ec2.rb (hunk 1 above)
Comment (domcleal) [inline app/models/compute_resources/foreman/model/ec2.rb]: `args[:security_group_ids].present?` should show if VPC is in use

Comment (haf) [inline app/models/compute_resources/foreman/model/ec2.rb]: ok, so despite security_groups being present in plain old EC2, too, them being present in this args hash shows the presence of vpc?

Comment (domcleal) [inline app/models/compute_resources/foreman/model/ec2.rb]: Ah yeah, I thought this attr was only used when VPCs are in effect, but perhaps not.

**Comment by haf** [comment]
What's wrong with my text now?

**Comment by domcleal** [comment]
@haf just "fixes #5999 - ....", it's one or the other :)

**Comment by dLobatog** [comment]
[test]

**Comment by haf** [comment]
ok, so if I could log into my dev env, I could test it too

**Comment by haf** [comment]
screen shot 2014-11-14 at 10 50 00

File: app/models/compute_resources/foreman/model/ec2.rb (hunk 1 above)
Comment (domcleal) [inline app/models/compute_resources/foreman/model/ec2.rb]: `args[:subnet_id].present?` might be more accurate as that's the field I think you select the VPC in, blank being none.

**Comment by domcleal** [comment]
Try `rake permissions:reset`

**Comment by dLobatog** [comment]
@haf the comment I made (test in square brackets) just lets Jenkins know it should test this PR, :smile:

File: .gitignore [hunk 2]
```diff
@@ -36,3 +36,4 @@ _build
 zeus.json
```
Comment (dLobatog) [inline .gitignore:36]: No need to change the .gitignore, please remove it from the commit.

File: app/models/compute_resources/foreman/model/ec2.rb [hunk 3]
```diff
@@ -103,6 +104,12 @@ def user_data_supported?
 
     private
 
+    # gets whether the args array is configured in a way that implies that we're
+    # in a VPC
+    def is_vpc args
```
Comment (dLobatog) [inline app/models/compute_resources/foreman/model/ec2.rb]: `def is_vpc?` makes the method a bit more idiomatic. The comment here is a bit redundant given the method name already tells what it does in my opinion.

**Comment by dLobatog** [comment]
@haf Nice, tests are :green_apple: , so if you don't mind recommitting with the changes mentioned above, I'll merge after that! Thanks!

**Comment by haf** [comment]
How can this work?

```
irb(main):001:0> {}[:alal].present?
NoMethodError: undefined method `present?' for nil:NilClass
    from (irb):1
    from /Users/henrik.feldt/.rbenv/versions/2.1.2/bin/irb:11:in `<main>'
```

**Comment by dLobatog** [comment]
@haf It looks like you're trying to run `.present?` outside of Rails, in plain Ruby. Try running it in a Rails console and you'll see it works.

[test]

**Comment by domcleal** [comment]
@haf Rails adds .present? and various other methods to core objects.  It helps with consistency a bit.  Try using `rails c` (or `foreman-rake console` on a prod install) and test there instead of using irb.

<pre>
2.0.0-p353 :001 > {}[:alal].present?
 => false 
</pre>

**Comment by haf** [comment]
lol @ rails and extending NilObject :)

**Comment by dLobatog** [comment]
[test]

**Comment by haf** [comment]
Login is still now working after `rake permissions:reset`

**Comment by haf** [comment]
Test is successful

**Comment by haf** [comment]
undefined local variable or method `vm' for #Foreman::Model::EC2:0x00000008db8ca8

**Comment by haf** [comment]
I don't always test my code, but when I do, I do it in production.

I just tested this PR in production. Consider it tested and ready to merge.

**Comment by haf** [comment]
When using this PR it still waits on the public IP: `Starting SSH provisioning script - waiting for 54.77.xxx.xxx to respond`

**Comment by haf** [comment]
Also the method for selecting security groups is buggy, it defaults to four empty values and still allows one to click through when provisioning in EC2.

**Comment by haf** [comment]
Also foreman blocks the web thread waiting for SSH to respond, instead of polling a "current status" endpoint of the like.
screen shot 2014-11-14 at 15 35 59

**Comment by haf** [comment]
Is there a way to tell foreman to use the private IP and not the public IP (that we only use for NAT) when SSHing into the instance?

**Comment by dLobatog** [comment]
[test]

**Comment by dLobatog** [comment]
@haf It is possible, but I think defaulting to the public IP is safer. Check [ssh_provision](https://github.com/theforeman/foreman/blob/develop/app/models/concerns/orchestration/ssh_provision.rb#L55) for that, it's where Foreman orchestration waits for the public IP to become online.

Of course you can file a new issue in Redmine if you think using the private IP would be much more useful in your case :smile: 

Code for this PR is fine in my opinion, although a test would be good. I'm having problems with my EC2 account, as soon as I recover it I'll test it and hopefully merge :+1: thanks again!

**Comment by dLobatog** [comment]
@haf Finally I got to test this, my ec2 account was completely borked. I'll merge in 3..2..1.. :bomb:

**Comment by dLobatog** [comment]
Merged as 5ba6c393e0a9bde31ffda6e10a783d3af21c1143 , thanks @haf , and sorry for the delay. Looking forward for more contributions like this from you!

**Comment by haf** [comment]
=)

**Pull request closed without merging**